modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ * ```
+ *
+ * We can now [traverse] structures that contain strings parsing them into integers and either with a `fail-fast` strategy using `Either` or accumulating failures with `ValidatedNel`.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.extensions.either.applicative.applicative
+ * import arrow.core.k
+ * import arrow.core.ValidatedNel
+ * import arrow.core.extensions.either.foldable.isEmpty
+ * import arrow.core.invalidNel
+ * import arrow.core.validNel
+ *
+ * fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val list = listOf("1", "2", "3").k().traverse(Either.applicative(), ::parseIntEither)
+ *   val failFastList = listOf("1", "abc", "3", "4s").k().traverse(Either.applicative(), ::parseIntEither)
+ *   //sampleEnd
+ *   println("list= $list")
+ *   println("failFastList= $failFastList")
+ * }
+ * ```
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.Nel
+ * import arrow.core.ValidatedNel
+ * import arrow.core.extensions.nonemptylist.semigroup.semigroup
+ * import arrow.core.extensions.validated.applicative.applicative
+ * import arrow.core.fix
+ * import arrow.core.invalidNel
+ * import arrow.core.k
+ * import arrow.core.validNel
+ *
+ * fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val validatedList = listOf("1", "22w", "3", "33s").k()
+ *     .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
+ *   //sampleEnd
+ *   println(validatedList)
+ * }
+ * ```
+ *
+ * Notice that in the `Either` case, should any string fail to parse the entire [traverse] is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of [traverse] is closely tied with the [Applicative] behavior of the data type, where computations are runed in isolation.
```

Comment:
```suggestion
 * Notice that in the `Either` case, should any string fail to parse the entire [traverse] is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of [traverse] is closely tied with the [Applicative] behavior of the data type, where computations are run in isolation.
```

This a good place to introduce three other concepts as well: 
`traverse` order (which depends on `Foldable` but is generally always left to right for most of arrows datatypes).
And also laziness, `traverse` (with an applicative that supports it, by implementing `lazyAp`) will not perform unnecessary steps and work on infinite structures.
And lasty the evaluation order of your applicative matters: For all arrow types it is left to right (listOf(1,2,3,4).traverse { IO { println(it); it } } puts out 1,2,3,4 and returns the list in the same order ) but you can easily define a type `Backwards` that executes the effects in reverse order (by flipping argument order of ap). Backwards would still return the list in same order but would print 4,3,2,1.

Not sure which of these concepts are important enough to be included, your choice :), but this could be a good place for it